Sources/NIOConcurrencyHelpers/fastatomics.swift

```diff
@@ -0,0 +1,269 @@
+    /// This implementation conforms to C11's `atomic_compare_exchange_strong`. This
+    /// means that the compare-and-swap will always succeed if `expected` is equal to
+    /// value. Additionally, it uses a *sequentially consistent ordering*. For more
+    /// details on atomic memory models, check the documentation for C11's
+    /// `stdatomic.h`.
+    ///
+    /// - Parameter expected: The value that this object must currently hold for the
+    ///     compare-and-swap to succeed.
+    /// - Parameter desired: The new value that this object will hold if the compare
+    ///     succeeds.
+    /// - Returns: `True` if the exchange occurred, or `False` if `expected` did not
+    ///     match the current value and so no exchange occurred.
+    @inlinable
+    public func compareAndExchange(expected: T, desired: T) -> Bool {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_compare_and_exchange($0, expected, desired)
+        }
+    }
+
+    /// Atomically adds `rhs` to this object.
+    ///
+    /// This implementation uses a *relaxed* memory ordering. This guarantees nothing
+    /// more than that this operation is atomic: there is no guarantee that any other
+    /// event will be ordered before or after this one.
+    ///
+    /// - Parameter rhs: The value to add to this object.
+    /// - Returns: The previous value of this object, before the addition occurred.
+    @inlinable
+    public func add(_ rhs: T) -> T {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_add($0, rhs)
+        }
+    }
+
+    /// Atomically subtracts `rhs` from this object.
+    ///
+    /// This implementation uses a *relaxed* memory ordering. This guarantees nothing
+    /// more than that this operation is atomic: there is no guarantee that any other
+    /// event will be ordered before or after this one.
+    ///
+    /// - Parameter rhs: The value to subtract from this object.
+    /// - Returns: The previous value of this object, before the subtraction occurred.
+    @inlinable
+    public func sub(_ rhs: T) -> T {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_sub($0, rhs)
+        }
+    }
+
+    /// Atomically exchanges `value` for the current value of this object.
+    ///
+    /// This implementation uses a *relaxed* memory ordering. This guarantees nothing
+    /// more than that this operation is atomic: there is no guarantee that any other
+    /// event will be ordered before or after this one.
+    ///
+    /// - Parameter value: The new value to set this object to.
+    /// - Returns: The value previously held by this object.
+    @inlinable
+    public func exchange(with value: T) -> T {
+        return Manager(unsafeBufferObject: self).withUnsafeMutablePointerToElements {
+            return T.fast_atomic_exchange($0, value)
```

Comment:
@weissi @Lukasa What's your stance on implicit returns? Both returns in each of these functions can be elided but I included them since I don't know the preferred style.

Reply:
Mixed. We tend to allow them within single-statement closures, but as we support Swift 5 we forbid them everywhere else.